Single Step Blocks... sort of!

eddieantonio opened the pull request:
Request for comments on the design of the stuff I have. Specifically, I want your comments on how I'll integrate context blocks into this, since the `Context` class _needs_ to call every `block.run()` whereas context blocks do their own calling of `block.run()` which is why everything still inexplicably works.

Would fix #1059 and maybe even #1057 and probably #1061.

eddieantonio commented:
> **WARNING**: Long, probably very confusing post. Please ask questions!

There are currently 11 `<wb-context>` blocks defined: `control.whenProgramRuns`, `control.eachFrame`, `control.loopOver`, `control.broadcast`, `control.receive`, `control.if`, `control.ifElse`, `control.comment`, `input.whenKeyPressed`, `motion.whenDeviceTurned`, `geolocation.whenLocationUpdated`.

My debugger _really_ wants to be in charge of calling `block.run(scope)`, but, as all of these context scripts demonstrate, that is up to the script to call. This is bad for my debugger!

So my thoughts are, for the scripts in `runtime.js`, we (probably I) should rewrite them ALL to allow the debugger/execution model to call that precious `block.run()` method. So, instead of having:

```javascript
global.runtime = {
    control: {
        ifElse: function(args, containers){
            var self = this;
            if (args[0]){
                containers[0].forEach(function(block){
                    block.run(self); // EVIL!
                });
            }else{
                containers[1].forEach(function(block){
                    block.run(self); // EVIIIILLL!!!
                });
            }
        },
        /* ... */
    },
    /* ... */

};

```

Have one more level of nesting with a series of callbacks. The only required callbacks are `next()` and `setup()`. Note the `thread` argument, which is the current thread (as of now named `Context` in `execution.js` -- but will be renamed to not conflict with the name of context blocks). The thread ultimately controls what blocks get executed. It's up to the `setup()` callback to tell the thread what is the first block to execute. 

```javascript
global.runtime = {
    control:
        ifElse: {
            /*
             * Called by the runner every time it wants the next block
             * that should be run.
             *
             * `this` is bound to the "state" object of the current
             * frame.
             *
             * Returns null if there are no more blocks left to run.
             */
            next: function (thread, args, containers, elem) {
                /* Since `elem` is a <wb-step> or <wb-block>, this will
                 * always be the _next_ block to execute, or null once
                 * there are no more steps left in the container. */
                return elem.nextSiblingElement;
            },
            /**
             * Called with all of the values evaluated in `args` and all
             * of the contains belonging to the element.
             *
             * This callback should call `thread.newScope`, 
             * `thread.newFrame`, or `thread.newThread` to begin
             * running in a new scope and stack frame, or simply a new
             * frame with the same scope, or an entirely new thread
             * (like for asynchronous callbacks).
             *
             * `this` is bound to the "state" object of the current
             * frame.
             */
            begin: function (thread, args, containers, elem) {
                if (args[0]) {
                    // Use the "if" branch.
                    this.branch = container[0];
                } else {
                    // Use the "else" branch.
                    this.branch = container[1];
                }
                /* Start executing the if or else branch. */
                thread.newScope(this.branch);
            },
            /*
             * Called before values are evaluated (thus args is not
             * provided).
             * Returns false to prevent evaluation of values.
             */
            beforeScript: function (thread, containers, elem) {
                return true;
            }
        },
        /* ... */
    },
    /* ... */
};
```

eddieantonio commented:
By the way, `control.comment` is beautiful using this approach:

```javascript
global.runtime = {
    control:
        comment: {
            /* This should never get called, but just in case, it returns null -- no next instruction. */
            next: function (thread, args, containers, elem) {  return null; },
            /* Never calls `thread.*`, thus is a no-op.  */
            begin: function (thread, args, containers, elem) { },
            /* Do not evaluate arguments. Ever. */
            beforeScript: function (thread, containers, elem) {
                return false;
            }
        },
        /* ... */
    },
    /* ... */
};

dethe commented:
In principle I think this is the right way to go. Pay attention to patterns that arise while implementing to see if there are things we can do to simplify it, because I worry about the complexity.

I'd rename "thread" because it is confusing with multithreading, JavaScript is single-threaded until we start messing about with WebWorkers, etc.

caleybrock commented:
How does this effect that state diagram you showed us?

So if a user is using the debugger and they are working with a script that has `input.whenKeyPressed`, `motion.whenDeviceTurned`, `geolocation.whenLocationUpdated` and these get triggered when the user is in pause or something, then what happens?

caleybrock commented on js/execution.js:
```diff
@@ -0,0 +1,249 @@
+
+    /**
+     * Starts (asynchronous!) execution from scratch. Can only be called once.
+     */
```
why does this get called only once? Wouldn't you want to start from scratch at some point? For example, user hits stop, then user hits play.

eddieantonio replied:
Good point. Yeah, all of this non-sense here is so that each `WaterbearProcess` is a use-once only object. Every time start is pressed, it actually _tosses out_ any existing `WaterbearProcess` instance, and instantiates a new one. I decided this was a good idea because the process instance stores _a lot of state_, and this way I can throw it all out all in one go and start from scratch every time `start` is pressed. As noted in #998, we still have an issue with storing state from previous runs... somewhere-- and I wanted to avoid that altogether.

But I agree, it does look weird.

caleybrock replied:
Cool - makes sense

eddieantonio commented:
> So if a user is using the debugger and they are working with a script that has input.whenKeyPressed, motion.whenDeviceTurned, geolocation.whenLocationUpdated and these get triggered when the user is in pause or something, then what happens?

I honestly do know what _should_ happen. But the reason that I have this `Context` class (which should be renamed to something that is not `Thread`) is to encapsulate that asynchronous behaviour in a different execution context. And then the debugger code can chose when _exactly_ that code is run. The code that handles switching between execution contexts is supposed to go in `Process#nextStep` but it only handles one (the "root context").

JarrettSpiker commented:
Alright, just so I can get my head around this, you want to have something like a list of blocks to run, and then rewrite the scripts for ALL the blocks so that  so that they have a run function? The will get run by going through them in turn? All the context blocks would be able to change the list, so that different blocks are run depending on the context?

Am I way off base here?

dethe commented:
Don't all the blocks have a run function now?

I think it would keep thing more consistent if iterating through the contents of a context block was not particularly special, trying to keep every call to a block roughly the same. What would be ideal is if we can call `run()` on a block and then get notified when it is done, so we can call `run()` on the next block. That way, we can have any arbitrary pause between each block.

Does that make sense?

eddieantonio commented:
> Alright, just so I can get my head around this, you want to have something like a list of blocks to run, and then rewrite the scripts for ALL the blocks so that so that they have a run function? The will get run by going through them in turn? All the context blocks would be able to change the list, so that different blocks are run depending on the context?

All of the actual blocks -- `<wb-step>`, `<wb-expression>`, `<wb-step>` _already_ have a run method. Try it! Open up the console in the playground.html and do this:

```js
var b = document.querySelector('wb-step[script="control.log"]);
b.run({});
```

It should log... nothing! But it'll work.

But who calls the run method? I want that to _always_ be my weird executing thingy. Context blocks, however, call run as part of their script. See for instance my earlier example:

```js
global.runtime = {
    control: {
        ifElse: function(args, containers){
            var self = this;
            if (args[0]){
                containers[0].forEach(function(block){
                    // NOOOOOO
                    block.run(self);
                });
            }else{
                containers[1].forEach(function(block){
                    // NOOOOOOOOOOOOOOOOO!
                    block.run(self)
                });
            }
        },
        /* ... */
    },
    /* ... */

};
```

I don't want a list _per se_. I want each block to tell me what the _next_ block that should run is. Hence my `next()` callback that I proposed. But as with Dethe, I worry about the complexity of this approach.

eddieantonio commented:
> I think it would keep thing more consistent if iterating through the contents of a context block was not particularly special, trying to keep every call to a block roughly the same. What would be ideal is if we can call run() on a block and then get notified when it is done, so we can call run() on the next block. That way, we can have any arbitrary pause between each block.

So you want the calls to _run()_ to be asynchronous, calling a callback when they're done, so that the we can `run()` the next block?

eddieantonio commented:
# Progress update!

So this pull request is quickly ballooning in size. To cut it down a bit, I've decided to "properly" handling `each frame` in a different issue/PR. Similarly, things that depend on asynchronous events are left for another PR. Until then, here's what I have done:

 - Changed the API from taking multiple callbacks to one (as it was previously).
 - Changed name of `Context` to `Strand`. Strand are like Waterbear's threads!
 - Introduced concept of `Frame`. It turned out to be pretty important.
 - Converted some callbacks to the new API
 - Switch handling of scope to `Process`/`Strand` (though maybe it should really be in `Frame`?)

## Context callbacks converted to new API

 - [x] control.whenProgramRuns
 - [x] control.eachFrame
 - [ ] control.loopOver
 - [ ] control.broadcast
 - [ ]  control.receive
 - [x] control.if
 - [x]  control.ifElse
 - [x] control.commentContext
 - [ ] input.whenKeyPressed
 - [ ] motion.whenDeviceTurned,
 - [ ] geolocation.whenLocationUpdated

## Still left to do for this PR

 - [ ] Hook-up UI for pause/step
 - [ ] Styles for blocks that are paused
 - [ ] Merge from master, fixing conflicts
 - [ ] Converting `loop over` to new API'

caleybrock commented:
Very organized and putting things in separate PRs is even better, I approve :+1:

eddieantonio commented:
So I finally consider this PR complete. I've moved a bunch of things out of it, because it was already getting too large to be manageable. Review and merge when ready!
